### ISSUE TITLE
Investigate for and while loops as syntactic sugar

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#3820]
We should investigate adding syntactic sugar for two types of for-loops: An imperative-style loop and a monadic style-loop.

# Tasks
- [x] #3950
- [x] #3951

# Imperative Style

Add something like:

```scala
for (pat <- exp) exp_body
```

where `exp` must conform to `ForEach`.

which can be generalized to:

```scala
for (pat1 <- exp1; pat2 <- exp2; ...) exp
```

# Monadic Style

```scala
for (pat <- exp) yield exp_body
for (pat <- exp; if exp_cond) yield exp_body
```

where `exp` must conform to `FunctorFilter` or maybe `MonadFilter`?

# Open Questions:

- How will this interact with iterators, mutable collections, and regions?
- When/where to desugar?

**Comment by magnus-madsen** [comment]
Use https://github.com/paulbutcher/chess-flix/blob/master/src/Main.flix to ensure we have what we want.

**Comment by paulbutcher** [comment]
My immediate thought is that we should be able to pretty much lift exactly what Scala does verbatim? This is the best explanation of Scala's behaviour I'm aware of:

https://stackoverflow.com/questions/1052476/what-is-scalas-yield/1059501#1059501

It supports both imperative and monadic styles through very simple rewriting rules. I can't immediately see any reason why they (the Scala 2.7 version without `withFilter`) wouldn't work pretty much as-is in Flix? The answer to questions about how it would interact with iterators, mutable collections and regions are exactly the same as the answer for how the rewritten code (i.e. some combination of calls to `foreach`, `map`, `flatMap`, and `filter`) would work. Or am I missing something?

One possible enhancement over Scala would be to disallow the monadic style for anything that wasn't `Pure`?

As for when to desugar, how about a new desugar phase which would run very early in the sequence. Possibly right after the parser?

**Comment by paulbutcher** [comment]
The chess code referenced above would look like this:

```
for (
    solution <- solve(rows, cols, rest);
    x <- Set.range(0, rows);
    y <- Set.range(0, cols);
    let pos = (x, y);
    if allowed(piece, pos, solution)
) yield Set.insert((piece, pos), solution)
```

and would desugar to:

```
Set.flatMap(solution ->
    Set.flatMap(x ->
            Set.map(pos -> Set.insert((piece, pos), solution),
            Set.filter(pos -> allowed(piece, pos, solution),
                Set.map(y ->
                    (x, y),
                    Set.range(0, cols)))),
        Set.range(0, rows)),
    solve(rows, cols, rest))
```

**Comment by paulbutcher** [comment]
As the underscores show in the above, there will need to be something in there to handle unused variables. This is also an issue for Scala:

https://github.com/scala/bug/issues/10287

**Comment by magnus-madsen** [comment]
> It supports both imperative and monadic styles through very simple rewriting rules. I can't immediately see any reason why they (the Scala 2.7 version without `withFilter`) wouldn't work pretty much as-is in Flix? The answer to questions about how it would interact with iterators, mutable collections and regions are exactly the same as the answer for how the rewritten code (i.e. some combination of calls to `foreach`, `map`, `flatMap`, and `filter`) would work. Or am I missing something?

That's exactly right and what we would do. The questions are:

- Where does `filter`, `map`, and `flatMap` come from? 
- How do we deal with regions / effects.

For (1), presumably we don't want to import these three functions from various namespaces. Instead, we probably want to use the some type class. But which? `map` belongs to `Functor`, `flatMap` to `Monad` (which extends `Functor`). There is no type class for `filter` -- not sure exactly where that should come from. Anyway, this suggests that we would probably use the `Monad` type class. But `Set` and `Map` are not Monads so that would suck.

For (2), we have this challenge that mutable data structures are not `Functor`s or `Monad`s because they cannot conform to these interfaces because they carry an region parameter. (This is a bit abstract, but there are other tickets that discuss this particular issue).

Thus, again, we seem stuck at: "We need map, flatMap, filter" but where do we get them from?

**Comment by magnus-madsen** [comment]
(Note: `Set` has map and flatMap, but it is slightly dubious if it should.)

**Comment by magnus-madsen** [comment]
Another option is to come up with some Iterable/Iterator abstraction, which has also been discussed in other tickets, but there we are also a bit stuck.

**Comment by magnus-madsen** [comment]
TLDR: We need to think more, come up with a concrete proposal, and then hash it out :)

**Comment by paulbutcher** [comment]
https://github.com/flix/flix/issues/3950 and https://github.com/flix/flix/issues/3951 both say "some issues remain". Out of interest, what are those issues (and are they tracked anywhere)?

**Comment by magnus-madsen** [comment]
Open issues:

- (foreach) How can we make integrate both immutable and mutable collections into `foreach`. Currently for a mutable collection you can write `foreach(x <- m)` but for an immutable you must write `foreach(x <- iterator(r, m)`. 
- (for-yield) How can we generalize `for-yield` to work with different collections? Do we want to use MonadEmpty? What about Stephen's MonadZip how does that interact?

I should probably open tickets for both. In both cases we know what we want, but we don't know how or if it is even possible.

**Comment by magnus-madsen** [comment]
We have some open issues, but those can be addressed in separate tickets.
